docs/design/proactive-diagnosis.md

```diff
@@ -0,0 +1,113 @@
+# Proactive Diagnosis to Enhance Training Fault Tolerance
+
+## Background
+
+If you have already used DLRover or read any related introductions, you would 
+know that one important role of DLRover is to enhance the availability of 
+training tasks, meaning it can automatically handle faults and resume training 
+when encountering occasional exceptions. In the initial design, fault tolerance 
+during training was performed passively, meaning that only when a runtime 
+process or container failed and exited would the DLRover Master be notified and 
+attempt the corresponding fault-tolerant actions. However, as production scale 
+gradually increased, it became apparent that relying solely on passive fault 
+tolerance could not cover all exception scenarios. Therefore, the first 
+positive fault tolerance mechanism: 'the heartbeat mechanism' was implemented. 
+But considering that in distributed training, the training agent side also 
+employs a multi-process implementation, the aforementioned single heartbeat 
+mechanism cannot cover various exception scenarios that occur during actual 
+training operations (it can only cover some common cases).
+
+Therefore, a comprehensive, scalable framework is needed to gradually improve 
+the diagnosis and handling of corresponding exception scenarios.
+
+## Target
+
+- Be capable of proactive probing.
+- Requires flexible and scalable implementation.
+- Be able to reuse the current fault tolerance mechanisms.
```

Comment:
Can you list which failure we can diagnose?

Reply:
The currently implemented active diagnosis:
1.  heartbeat timeout 
2. training hang. 

Plan to refactor the previously implemented resource monitor into this framework, adding new diagnostic logic based on resource metrics.

Reply:
Doc updated.